test/hotspot/jtreg/containers/cgroup/NestedCgroup.java

```diff
@@ -0,0 +1,268 @@
+            cgcreate2.add(CONTROLLERS_PATH_INNER);
+            pSystem(cgcreate2, "", "", "", "", "");
+
+            String provider = Metrics.systemMetrics().getProvider();
+            System.err.println("Metrics.systemMetrics().getProvider() = " + provider);
+            boolean isCgroup2;
+            if ("cgroupv1".equals(provider)) {
+              isCgroup2 = false;
+            } else if ("cgroupv2".equals(provider)) {
+              isCgroup2 = true;
+            } else {
+              throw new IllegalArgumentException();
+            }
+            System.err.println("isCgroup2 = " + isCgroup2);
+
+            String mountInfo;
+            try {
+                mountInfo = Files.readString(Path.of(MOUNTINFO));
+            } catch (NoSuchFileException e) {
+                throw new SkippedException("Cannot open " + MOUNTINFO);
+            }
+
+            Matcher matcher = Pattern.compile("^(?:\\S+\\s+){4}(\\S+)\\s.*\\scgroup(?:2(?:\\s+\\S+){2}|\\s+\\S+\\s+(?:\\S*,)?memory(?:,\\S*)?)$", Pattern.MULTILINE).matcher(mountInfo);
+            if (!matcher.find()) {
+                System.err.println(mountInfo);
+                throw new SkippedException("cgroup/cgroup2 filesystem mount point not found");
+            }
+            sysFsCgroup = matcher.group(1);
+            System.err.println("sysFsCgroup = " + sysFsCgroup);
+
+            System.err.println(LINE_DELIM + " " + (isCgroup2 ? "cgroup2" : "cgroup1") + " mount point: " + sysFsCgroup);
+            memory_max_filename = isCgroup2 ? "memory.max" : "memory.limit_in_bytes";
+            Files.writeString(Path.of(sysFsCgroup + "/" + CGROUP_OUTER + "/" + memory_max_filename), "" + MEMORY_MAX_OUTER);
+
+            // Here starts a copy of ProcessTools.createJavaProcessBuilder.
+            List<String> cgexec = new ArrayList<>();
+            Limits limits = hook(cgexec);
+            OutputAnalyzer output = pSystem(cgexec);
+            // C++ CgroupController
+            output.shouldMatch("\\[trace\\]\\[os,container\\] Memory Limit is: " + limits.integer + "$");
+
+            pSystem(cgdelete);
+        }
+
+        public abstract Limits hook(List<String> cgexec) throws IOException;
+    }
+    private static class TestTwoLimits extends Test {
+        public Limits hook(List<String> cgexec) throws IOException {
+            // CgroupV1Subsystem::read_memory_limit_in_bytes considered hierarchical_memory_limit only when inner memory.limit_in_bytes is unlimited.
+            Files.writeString(Path.of(sysFsCgroup + "/" + CGROUP_OUTER + "/" + CGROUP_INNER + "/" + memory_max_filename), "" + MEMORY_MAX_INNER);
+
+            args_add_cgexec(cgexec);
+            args_add_self_verbose(cgexec);
+            cgexec.add("-version");
+
+            // KFAIL - verify the CgroupSubsystem::initialize_hierarchy() and jdk.internal.platform.CgroupSubsystem.initializeHierarchy() bug
+            // TestTwoLimits does not see the lower MEMORY_MAX_OUTER limit.
```

Comment:
Remove this obsolete(?) comment.

Reply:
This comment is documenting what the TestTwoLimits testcase does. Which I find useful.
It is KFAIL - Known Failure - the current OpenJDK code does not properly simulate what Linux kernel does. If you do:
```
cgset -r memory.max=$[1024*1024*1024] a/b
cgset -r memory.max=$[512*1024] a
cgexec -g memory:a/b java...
```
Then OpenJDK thinks it is `1024*1024*1024 KB` but Linux kernel will limit OpenJDK to `512*1024 KB`. So this problem is documented and tested.